### PR TITLE
[SPARK-12063][SQL] Use number in group by clause to refer to columns

### PR DESCRIPTION
**Pull request by dereksabryfb**
If there is a number n in a group by clause, the nth column in the select clause is used to modify the group by clause to refer to this column instead of the number.

eg.
select a,b from c group by 1,2
becomes
select a,b from c group by a,b

### CODE REVIEW

**Comment by marmbrus** [comment]
Thanks for working on this.  This seems resonable to support, but I have two suggestions:
 - I would probably implement this as a rule in the [`Analyzer`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala) so that it is not specific to the Hive parser.
 - Please add a unit test, probably in `SQLQuerySuite`

**Comment by rxin** [comment]
@dereksabryfb   you should also add the email you used in your git commit to your github profile so it shows up on github.

**Comment by dereksabryfb** [comment]
Thanks for your feedback! The change is now implemented as a rule in the Analyzer and a unit test has been added. Let me know if there are any other changes I need to make.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -178,6 +179,27 @@ class Analyzer(
...
+  /**
+   * Replaces queries of the form "SELECT expr FROM A GROUP BY 1"
+   * with a query of the form "SELECT expr FROM A GROUP BY expr"
+   */
+  object ResolveGroupByReferences extends Rule[LogicalPlan] {
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        Aggregate(groups.map(group => group match {
+          case g if g.prettyString forall Character.isDigit =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I'm not sure we want to match on the pretty string as I think this would also trigger for things like `"1"`, instead I'd consider matching on a Literal of NumericType.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -178,6 +179,27 @@ class Analyzer(
...
 
+  /**
+   * Replaces queries of the form "SELECT expr FROM A GROUP BY 1"
+   * with a query of the form "SELECT expr FROM A GROUP BY expr"
+   */
+  object ResolveGroupByReferences extends Rule[LogicalPlan] {
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        Aggregate(groups.map(group => group match {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Nit: would you mind assigning this to an intermediate variable.  this is a little hard to parse visually.

**Comment by marmbrus** [comment]
Are there other cases that we should handle here?  Can you do the same thing in ORDER BY normally?

**Comment by dereksabryfb** [comment]
Thanks for the feedback! I'm making the changes you suggested. With respect to the 'ORDER BY' clause, this looks to be semi-handled by the ResolveSortReferences in the Analyzer; because in standard SQL literals are allowed in the order by clause, this rule just sorts by the literal value '1'; In HiveQL, '1' is interpreted as a column in the same way it is in the group by clause.

I can add the case for a Sort() with an IntegerType. I assume there is no one currently relying on how 'sort by 1' currently functions.

**Comment by marmbrus** [comment]
We'll have to note the change in the release notes, but since its a no-op to sort by a constant I think we can safely change behavior here.

**Comment by dereksabryfb** [comment]
Added a case for sort

**Comment by marmbrus** [comment]
ok to test

**Comment by dereksabryfb** [comment]
Apologies, I haven't been able to run ./dev/run-tests, getting the following exception: http://pastebin.com/L0p0sjtJ

so I wasn't able to pick up the style issues, and I'm not sure if there's more that the build doesn't flag.

**Comment by marmbrus** [comment]
I'd try the following locally `build/sbt scalastyle test:scalastyle catalyst/test sql/test`.

Each of those commands can be run separately too and you can use ~ to rerun whenever something changes to iterate more quickly `build/sbt ~scalastyle`

**Comment by dereksabryfb** [comment]
Looks like it fails on the query "SELECT a, count(2) FROM testData2 GROUP BY a, 2" because 2 now refers to the column count(2) which is not valid syntax for a group by clause in hive. How should I go about resolving this? The purpose of the test is to see that literals in the group by clause don't modify the results, but the purpose of this patch is to do the opposite of that. I could modify the offending case, but I feel the whole test may be irrelevant with this patch.

**Comment by marmbrus** [comment]
We should probably throw an `AnalysisException` for this if they use a column ordinal that refers to an aggregate expression.  The fact that we make it all the way to the execution is pretty confusing to a user.

Regarding the test, we can probably remove it.

**Comment by dereksabryfb** [comment]
I removed the offending test.

I found that there was no `AnalysisException` thrown even if there was an explicit aggregate in the group by clause (e.g. select a from b group by count(a)), and it would fail in the same way, so I added the check to `CheckAnalysis` ; if you think this is out of the scope of this pull request (since it isn't strictly to do with a number reference), I can create a new task and attach just that commit to it.

Thanks again for your feedback.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -178,6 +179,46 @@ class Analyzer(
...
+  /**
+   * Replaces queries of the form "SELECT expr FROM A GROUP BY 1 ORDER BY 1"
+   * with a query of the form "SELECT expr FROM A GROUP BY expr ORDER BY expr"
+   */
+  object ResolveNumericReferences extends Rule[LogicalPlan] {
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        val newGroups = groups.map {
+          case group if group.isInstanceOf[Literal] && group.dataType.isInstanceOf[IntegralType] =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: ```case Literal(index: Int) => ``` is easier. It also eliminates the need for ```group.toString.toInt```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -178,6 +179,46 @@ class Analyzer(
...
+                }
+              case a: Alias => a.child
+              case a: AttributeReference => a
+            }
+          case group => group
+        }
+        Aggregate(newGroups, aggs, child)
+      case Sort(ordering, global, child) =>
+        val newOrdering = ordering.map {
+          case o if o.child.isInstanceOf[Literal] && o.dataType.isInstanceOf[IntegralType] =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Same comment as above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -178,6 +179,46 @@ class Analyzer(
...
+            val newExpr = child.asInstanceOf[Project].projectList(o.child.toString.toInt - 1)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Tests are currently failing to due this cast. A ```Sort```'s child doesn't have to be a ```Project```, use the LogicalPlan ```output``` attributes to get the proper attribute.

**Comment by hvanhovell** [comment]
@dereksabryfb are you still working on this?

**Comment by srowen** [comment]
Ping @dereksabryfb -- are you working on this? or else close it

**Comment by hvanhovell** [comment]
I think this one can be closed, it has been implemented in https://github.com/apache/spark/pull/11846

**Comment by JoshRosen** [comment]
Yep, it looks like this PR has been subsumed by #11846. @dereksabryfb, could you please close this pull request? Thanks!
